narwhal/types/src/primary.rs

```diff
@@ -30,9 +31,61 @@ use std::{
 /// The round number.
 pub type Round = u64;
 
+/// The epoch UNIX timestamp in milliseconds
+pub type TimestampMs = u64;
+
+pub trait Timestamp {
+    fn elapsed(&self) -> Duration;
+}
+
+impl Timestamp for TimestampMs {
+    // Returns the time elapsed between the timestamp
+    // and "now". The result is a Duration.
+    fn elapsed(&self) -> Duration {
+        let diff = now().saturating_sub(*self);
+        Duration::from_millis(diff)
+    }
+}
+// Returns the current time expressed as UNIX
+// timestamp in milliseconds
+fn now() -> TimestampMs {
+    match SystemTime::now().duration_since(SystemTime::UNIX_EPOCH) {
+        Ok(n) => n.as_millis() as TimestampMs,
+        Err(_) => panic!("SystemTime before UNIX EPOCH!"),
+    }
+}
+
+// Additional metadata information for an entity. Those data
+// should not be treated as trustworthy data and should be used
+// for NON CRITICAL purposes only.
```

Comment:
what does "non critical" mean? Can you be more specific in this warning?

Reply:
Does it mean they are simply self-declared meta-data?

Reply:
For what do we use them?

Reply:
Basically metadata generated by the node it self which I don't believe we do have a way to verify (ex I can't verify for sure when exactly the certificate has been created - I do have to trust the proposer for it) . Now we are going to use them for analytics purposes. But in the future we can add anything there that we might not consider them "verifiable" but useful for other purposes (ex tracking/analytics).